storage/remote/client_test.go

```diff
@@ -11,7 +11,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package remote
+package remote_test
```

Comment:
So depends on context, but usually this is antipattern - I think on Prometheus we stick for tests to have exactly the same package. Can we revert this and same with exporting `recoverableError`?

Reply:
Yes, and that's standard Go practice AFAIK

Reply:
Actually it's easiest way to avoid `import cycle` (client test -> api -> client).

Also, why I can't receive your comments from email @bwplotka , really weird 😂

Reply:
No idea. So you were not notified about my review?

Reply:
Also yea, we need to fix the cycle in different way then.

Reply:
Could from Github page, but not email 😄 

If we want to reuse the package `web/api/v1`, then the cycle can not be avoided.

Maybe we could simulate an prometheus instance like thanos do ?

But it seems a big change, so we could do it in another PR (or several PR) ?

@bwplotka  WDYT?

Reply:
ping @bwplotka

Reply:
Sorry for delay. I looked through and we have 2 options:
* Mock server in some other way
* Move client code to separate package e.g `storage/remote/client`

I really like the second option because as far as I can see there is literally no import for `remote` stuff from `client.go` or `client_test.go` so it is actually might be more natural way.
Plus we can use exact api code for server side for both remote read and write :+1:  What do you think?

Reply:
`storage/remote ` is not using anything from `storage/remote/client/client.go` and `storage/remote/client/client_test.go` AFAIK  only `api/v1` is using this. 

> > Plus we can use exact api code for server side for both remote read and write +1 What do you think?
> Could you describe it in more detail? :)

Sure, I mean that the extra benefit for moving client to separate package is that we can in `api/v1_test.go` use exact server remote read code instead of just mocking something that *should* behave like server code, right?

Reply:
Yes, use exact api code for server side for both remote read and write is great.

But `storage/remote` surely use client related code for remote write and read 😂 

https://github.com/prometheus/prometheus/blob/master/storage/remote/storage.go#L72

https://github.com/prometheus/prometheus/blob/master/storage/remote/write.go#L135

Reply:
Yea, sorry, I missed that - it's quite tightly together. Still we have few options. I think making testing package external is not really preferable as it's a just hack for the potentially improper composition of packages. I would also try to avoid mocking server. 

In this case, I would say, let's just test client e2e in `api/v1_test.go` - let's not spend to much time on this!

The long term solution would be to allow `remote/storage.go` and `write.go` to work on `RemoteClient` interface rather and have `main` to put `remote/client` implementation (:

Reply:
Actually after more digging into the code, I find the remote part of `TestEndpoints` in `web/api/v1/api_test.go` almost do the same thing as this PR 😱 Maybe after adding ExternalLabels part, it would be enough for us.

Sad to spend so much time to do the duplicate thing 😂